### PR TITLE
Merge DMI and main repo

### PR DESCRIPTION
**Pull request by TillRasmussen**
Changes are all tested to be binary identical with an intel compiler. Build executable has also been carried out on hpc. These are not tested four coupled simulations with NUOPC_CESMBETA. Main purpose is to align DMI branch with main HYCOM.

CESMBETA/ocn_comp_nuopc_mod.F90

Call to HYCOM init: Removed restart_write option. Not used

blkdat.F90
Changed input to indicate whether to coldstart or restart
Dont read limits
Remove USE_NUOPC_CESMBETA

diapfl.F90 
Remove USE_NUOPC_CESMBETA

Mod_cb_arrays

sea ice import fields are now real (8)
      real(8), target, allocatable,dimension (:,:) :: &
             sic_import, & !Sea Ice Concentration
            sitx_import, & !Sea Ice X-Stress
            sity_import, & !Sea Ice Y-Stress
            siqs_import, & !Solar Heat Flux thru Ice to Ocean
            sifh_import, & !Ice Freezing/Melting Heat Flux
            sifs_import, & !Ice Freezing/Melting Salt Flux
            sifw_import, & !Ice Net Water Flux
             sit_import, & !Sea Ice Temperature
             sih_import, & !Sea Ice Thickness
             siu_import, & !Sea Ice X-Velocity
             siv_import    !Sea Ice Y-Velocity
Are these only used in CESMBETA directory files, ESMF and NAVY_coupled?
Maybe they should not be in the mod_cb_arrays

ocn_cpl_frq has also been changed to real(8). This is also only used in  CESMBETA directory files

Mod_hycom.F90
Read limits once
Simplified definition of HYCOM_init and HYCOM_Run
Introduced cpp flag DMI_CICE_COUPLED. Mainly to avoid coupling with atmosphere

mod_momtum.F90
Reduced number of USE_NUOPC_CESMBETA. The only remainder is error messages when wndflg has the wrong value

mod_restart:
Removed interface.

mxkprf.F90
removed  USE_NUOPC_CESMBETA

### CODE REVIEW

File: diapfl.F90 [hunk 1]
```diff
@@ -377,15 +377,12 @@ real fri                 ! function of Rig for KPP shear instability
...
 !
 ! --- do rivers here because difs is also used for tracers.
-#if defined (USE_NUOPC_CESMBETA)
       if(cpl_orivers.and.cpl_irivers) then
          riv_input = imp_orivers(i,j,1)+imp_irivers(i,j,1)
       else
          riv_input = rivers(i,j,1)
       endif
-#else
       riv_input = rivers(i,j,1)
```
Comment (abozec) [inline diapfl.F90:385]: line 385 needs to be removed for it to run as intended.

Comment (TillRasmussen) [reply diapfl.F90:385]: thanks, that is right. I will rerun my test and update

**Comment by abozec** [comment]
Few changes are needed. I've checked the code with CESM and ESPC and I manage also to get it bit for bit with the previous version, however it needs some changes to compile and run properly with both caps. 

The easiest change is in diapfl.F90, line 385 needs to be removed for it to work as intended.

In mod_cb_arrays.F90, line 837 #if ! defined (ESPC_COUPLE) needs to be removed since we are using that macro with the ESPC cap and we still need the declaration of the tml,sml etc arrays for the mod_restart.F90
However, doing that means that we now have a double declaration of umxl and vmxl when using the ESPC cap. I would suggest, you change umxl and vmxl into uml and vml  in:
mod_cb_arrays.F90
mod_hycom.F90 (leave the ones in Export_ESMF)
mod_restart.F90 
CESMBETA/mod_cb_arrays_nuopc_glue.F90
CESMBETA/mod_hycom_nuopc_glue.F90

**Comment by awallcraft** [comment]
real(8) does not necessarily imply real\*8, kind=8 might not be defined or might be real\*4 with some compilers.  So, for simplicity, HYCOM has continued to use real\*8 when 64-bit precision is required.  This is a "legacy" feature that isn't standard Fortran, but is universally supported by mainstream compilers.

In this case the _import arrays have to be real, because they can be arguments to communication routines that work on real arrays.  

mod_hycom.F90:      call xctila( sic_import,1,1,halo_ps)  !Sea Ice Concentration

In practice, real is real\*8 due to the compiler options we use but we reserve real\*8 for variables that *must* be real\*8.

**Comment by TillRasmussen** [comment]
I did not change the real*8 . These are linked to calls in the nuopc/esmf structure and fails if compiled on cray compiler.

**Comment by abozec** [comment]
Could you put all real(8) under the DMI macro ? e.g.:
#if defined (DMI_CICE_COUPLED)
       real(8), target, dimension (1-nbdy:idm+nbdy,1-nbdy:jdm+nbdy) :: &
#else
       real, target, dimension (1-nbdy:idm+nbdy,1-nbdy:jdm+nbdy) :: &
#endif
Then we should be able to merge.

**Comment by TillRasmussen** [comment]
The real (8) 2d arrays are no longer present in  mod_cb_array.F90. They have been made obsolete in the DMI code.

**Pull request merged**